lib/Artifact.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright © 2018 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    EventFired,
+    HandlerContext,
+    HandlerResult,
+    Success,
+} from "@atomist/automation-client";
+import { subscription } from "@atomist/automation-client/lib/graph/graphQL";
+import {
+    addressChannelsFor,
+    ArtifactGoal,
+    ArtifactListenerInvocation,
+    ArtifactListenerRegisterable,
+    DefaultGoalNameGenerator,
+    findSdmGoalOnCommit,
+    FulfillableGoalDetails,
+    FulfillableGoalWithRegistrations,
+    getGoalDefinitionFrom,
+    Goal,
+    logger,
+    PushListenerInvocation,
+    SdmGoalState,
+    SideEffect,
+    toArtifactListenerRegistration,
+    updateGoal,
+} from "@atomist/sdm";
+import * as _ from "lodash";
+import { OnImageLinked } from "./typings/types";
+
+export interface ArtifactRegistration {
+    listeners?: ArtifactListenerRegisterable[];
+}
+
+export class Artifact extends FulfillableGoalWithRegistrations<ArtifactRegistration> {
+    constructor(goalDetailsOrUniqueName: FulfillableGoalDetails | string
+                    = DefaultGoalNameGenerator.generateName("artifact"),
+                ...dependsOn: Goal[]) {
+
+        super({
+            ...ArtifactGoal.definition,
+            ...getGoalDefinitionFrom(goalDetailsOrUniqueName, DefaultGoalNameGenerator.generateName("artifact")),
+            displayName: "artifact",
+        }, ...dependsOn);
+
+        const fulfillment: SideEffect = {name: "build"};
+        this.addFulfillment(fulfillment);
+        this.sdm.addEvent({
```

Comment:
this direct access to `this.sdm` only works when the goal is created within the machine.ts file. Try it with `const artifact = new Artifact()` in a goals.ts file, and it will likely fail because this.sdm is undefined. 

To make this work, overwrite the `public register(sdm: SoftwareDeliveryMachine): void {` function of `FulfillableGoal` (make sure to call super.register(sdm) first) and add the event handler there.